spec/controllers/moderator_applications_controller_spec.rb

```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+# The application ID to use in requests.
+def application_id
+  # @application.id will always be set unless :does_not_exist is set for the example,
+  # in which case we specifically want to use an id that does not exist.
+  # The number 192_139 has no special significance.
+  @application.id || 192_139
+end
+
+def it_handles_nonexistent_application
+  it 'handles a non-existent application', :does_not_exist do
+    expect_not_found
+  end
+end
+
+def it_requires_that_applications_are_open
+  it 'requires that applications are open', :applications_closed do
+    expect_unauthorized
+  end
+end
+
+def it_requires_login
+  it 'requires login', :no_login do
+    expect_unauthenticated
+  end
+end
+
+def it_requires_correct_login
+  it_requires_login
+
+  it 'requires correct login', :incorrect_login do
+    expect_unauthorized
+  end
+end
+
+# The show and edit routes both have identical tests, which I have factored out here.
+def show_and_edit_tests(route)
+  before(:each) { get route, params: { id: application_id } }
+
+  it_handles_nonexistent_application
+  it_requires_that_applications_are_open
+  it_requires_correct_login
+
+  it 'allows admin access', :admin_login do
+    expect_successful_template route
+  end
+
+  it 'allows admin access even if applications are closed', :admin_login, :applications_closed do
+    expect_successful_template route
+  end
+
+  it 'works' do
+    expect_successful_template route
+  end
+end
+
+describe ModeratorApplicationsController do
+  before(:each) do |example|
+    ENV['MODERATOR_APPLICATIONS'] = example.metadata[:applications_closed] ? 'closed' : 'open'
```

Comment:
Probably should add a comment noting that 'closed' doesn't hold any significance over nil, just that it's cleaner to assign an actual value over assigning to nil (or whatever Ruby's syntax for that would be). I mean, I guess *technically* it would be prudent to verify that every case works, but when it gets down to it, I don't think anyone really cares that much.